kvpress/presses/kvzip_press.py

```diff
@@ -0,0 +1,358 @@
+
+        def wrapped_forward(*args, **kwargs):
+            self._context_ids = kwargs.get("input_ids", None)
+            self._cache = kwargs.get("past_key_values", None)
+            return original_forward(**kwargs)
+
+        model.forward = MethodType(wrapped_forward, model)
+
+        hooks = []
+        try:
+            yield
+            model.forward = original_forward  # Restore original
+
+            # After yield: KVzip scoring and compression phase
+            if self.compression_ratio > 0 and self._context_ids is not None:
+                # Now register attention hooks for compression
+                for layer in model.model.layers:
+                    if isinstance(model, Gemma3ForCausalLM) and layer.is_sliding:
+                        continue
+                    layer.self_attn.rotary_emb = model.model.rotary_emb
+                    hooks.append(layer.self_attn.register_forward_hook(self.forward_hook, with_kwargs=True))
+
+                self._perform_kvzip_compression(model, tokenizer)
+        finally:
+            for hook in hooks:
+                hook.remove()
+
+    def forward_hook(self, module: nn.Module, input: list[torch.Tensor], kwargs: dict, output: list):
+        """
+        Overwrite the forward_hook of BasePress
+        """
+
+        hidden_states = kwargs["hidden_states"]
+        cache = kwargs["past_key_value"]
+
+        if isinstance(cache, QuantizedCache):
+            keys = cache._dequantize(cache._quantized_key_cache[module.layer_idx])
+            values = cache._dequantize(cache._quantized_value_cache[module.layer_idx])
+        else:
+            keys = cache.key_cache[module.layer_idx]
+            values = cache.value_cache[module.layer_idx]
+
+        # Only do scoring. Compression is done afterward by press.compress_post(self.model) in pipeline.py.
+        keys, values = self.score(module, hidden_states, keys, values, output[1], kwargs)
+
+        if isinstance(cache, QuantizedCache):
+            cache._quantized_key_cache[module.layer_idx] = cache._quantize(keys, axis=cache.axis_key)
+            cache._quantized_value_cache[module.layer_idx] = cache._quantize(values, axis=cache.axis_value)
+            cache.key_cache[module.layer_idx] = torch.zeros(0, dtype=keys.dtype, device=keys.device)
+            cache.value_cache[module.layer_idx] = torch.zeros(0, dtype=keys.dtype, device=keys.device)
+            cache._seen_tokens = keys.shape[2]
+        else:
+            cache.key_cache[module.layer_idx] = keys
+            cache.value_cache[module.layer_idx] = values
+
+        return output
+
+    def _perform_kvzip_compression(self, model: PreTrainedModel, tokenizer: PreTrainedTokenizer):
+        """
+        Perform the KVzip scoring and compression algorithm.
+        """
+
+        # Prepare chunked inputs for context reconstruction
+        self.context_length = self._context_ids.shape[1]
+        input_ids = self.prepare(model, tokenizer)
```

Comment:
Here, and in some other occasions, please rename variables.
`input_ids` is a bit confusing, as input_ids usually equals the tokenized text ids.

E.g.
```
chunked_reconstruction_pairs = self.prepare(model, tokenizer)
```